File: Week_04/java/src/_81_SearchInRotatedSortedArrayII.java
/**
 * 第一遍：2020/09/13周日 ✅
 * 第二遍：2020/06/21周日
 * 第二遍：2020/06/12周二
 * 第三遍：2020/06/18周四
 * 第四遍：2020/07/02周四
 */
public class _81_SearchInRotatedSortedArrayII {
    static class BinarySearch {
        public boolean search(int[] nums, int target) {
            int l = 0, r = nums.length - 1;
            while (l <= r) {
                int m = (l + r) >>> 1;
                if (target == nums[m]) return true;
                if (nums[m] > nums[r]) {
                    if (target >= nums[l] && target < nums[m]) {
                        r = m - 1;
                    } else {
                        l = m + 1;
                    }
                } else if (nums[m] < nums[r]) {
                    if (target > nums[m] && target <= nums[r]) {
                        l = m + 1;
                    } else {
                        r = m - 1;
                    }
                } else {
                    if (nums[m] > nums[l]) {
                        if (target < nums[m]) {
                            r = m - 1;
                        } else {
                            l = m + 1;
                        }
                    } else if (nums[m] < nums[l]) {
                        r = m - 1;
                    } else {
                        r--;
                    }
                }
            }
            return false;
        }
    }
}

File: Week_04/c/src/_81_SearchInRotatedSortedArrayII.c
#include <stdbool.h>
bool search(int *nums, int n, int t)
{
    int l = 0, r = n - 1, m;
    while (l <= r)
    {
        m = l + ((r - l) >> 1);
        if (t == nums[m])
            return true;
        if (nums[m] > nums[l])
        {
            if (t >= nums[l] && t < nums[m])
            {
                r = m - 1;
            }
            else
            {
                l = m + 1;
            }
        }
        else if (nums[m] < nums[l])
        {
            if (t > nums[m] && t <= nums[r])
            {
                l = m + 1;
            }
            else
            {
                r = m - 1;
            }
        }
        else if (nums[m] != nums[r])
        {
            l = m + 1;
        }
        else
        {
            l++;
        }
    }
    return false;
}